addons/website_sale/models/product.py

```diff
@@ -170,6 +178,44 @@ def _compute_parents_and_self(self):
             else:
                 category.parents_and_self = category
 
+    def _get_categories_and_count_from_product_domain(self, categ_domain):
+        """ Return a tuple with the main product.public.category and a dict with category id as key and product count as value """
+        Product = self.env['product.template']
+        where_query = Product._where_calc(categ_domain)
+        Product._apply_ir_rules(where_query)
+        from_clause, where_clause, args = where_query.get_sql()
+        query = """
+            SELECT c.id, COUNT(DISTINCT product_template)
+            FROM %s
+            INNER JOIN %s ON product_template_id = product_template.id
+            INNER JOIN %s c ON product_public_category_id = c.id
+            WHERE %s
+            GROUP BY c.id
+        """ % (self._fields['product_tmpl_ids'].relation, Product._table, self._table, where_clause)
+        self.flush()
+        self.env.cr.execute(query, args)
+        categs_nbr_prod = dict(self.env.cr.fetchall())
+        categs_ids = list(categs_nbr_prod.keys())
```

Comment:
The result of `keys()` is already a list I think? You need to copy it for a reason?

Reply:
No it is a view object

Reply:
Ok, good to know!